### PR TITLE
upgrade mapbox-upload-limits [#92]

### PR DESCRIPTION
**Pull request by sumarlidason**
https://github.com/mapbox/mapbox-upload-validate/issues/92

* upgrade node ci tested versions to 6.x.x & 8.x.x    
* add test prerequisites
* add .npmrc to silence npm    
* AUDIT: upgrade some packages    
* improve module collision detection (npm ls)
* upgrade mock-fs from 3.8.0 -> 4.7.0 for node 8.x.x support
* fix bug with stream processing in serialtiles.js
* fix typo in package.json versions x.00.x -> x.0.x
* fix bug w/ STDOUT parsing w/ warnings enabled (tmpdir/tmpDir) (i think introduced by node 8.x.x)
* improve logging for bin.mapbox-upload-validate tests
* add package-lock.json

These updates will be used in atlas-ddb here,  where atlas-ddb reads and validates mbtiles for the atlas on-prem.: https://github.com/mapbox/atlas-ddb/compare/mapbox-upload-validate-test#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R34

### CODE REVIEW

**Comment by sumarlidason** [comment]
suggest bump to `5.0.0`

File: lib/validators/serialtiles.js [hunk 1]
```diff
@@ -23,25 +23,28 @@ function validateSerialtiles(opts, callback) {
...
     }
-    
     var read = fs.createReadStream(opts.filepath);
     read.pipe(zlib.createGunzip())
       .pipe(tilelive.deserialize())
       .once('error', fail)
       .pipe(validationStream)
       .once('error', fail)
-      .on('finish', callback).resume();
+      .on('finish', function() { if (!failed) callback(); })
```
[the remark this answers is not in the document]
Comment (sumarlidason) [reply lib/validators/serialtiles.js]: This bug surfaced when adding support for node 8.x.x from what I can tell. The behavior regarding pipes & errors must have changed.

This code from master on node v8.12.0:

```javascript
var read = fs.createReadStream(opts.filepath);
read.pipe(zlib.createGunzip())
  .pipe(tilelive.deserialize())
  .once('error', fail)
  .pipe(validationStream)
  .once('error', fail)
  .on('finish', callback).resume();
```

executes the `finish` callback twice and fails to report the error we'd like to see:

```bash
TAP version 13
# lib.validators.serialtiles: invalid gzipped file format
CALLBACK
ok 1 expected error
ok 2 expected error code
ok 3 expected error message
CALLBACK
not ok 4 expected error
not ok 5 plan != count
```


this phenomenon is described here, https://www.bennadel.com/blog/2692-you-have-to-explicitly-end-streams-after-pipes-break-in-node-js.htm

The refactor/fix is backwards compatible to 6.

## Code Coverage Report
codecoverage-enhancement_92

**Comment by sumarlidason** [comment]
bug: `queue-async.defer()` chain does not exit on error.

**Comment by sumarlidason** [comment]
- [x] moving to sequential workflow for validation only makes sense.
- [ ] also need to change validation method for `dataCounts()` function from `select count` to `select * limit 10` or similar as `select count()` with no `where` clause incurs significant IO[1] wait on the instance and prohibitively long[2] runtimes for validation.

[1]
```bash
avg-cpu:  %user   %nice %system %iowait  %steal   %idle
           0.20    0.00    0.10   49.55    0.10   50.05

Device:            tps    kB_read/s    kB_wrtn/s    kB_read    kB_wrtn
loop0             0.00         0.00         0.00          0          0
loop1             0.00         0.00         0.00          0          0
loop2             0.00         0.00         0.00          0          0
xvda            198.80     10988.00         0.00      54940          0
```

[2] will update this ticket once it finishes.

**Comment by sumarlidason** [comment]
Eval times for mbtiles `80gb` file:
```bash
16.72 user
45.48 system
1:14:48 elapsed (real)
```

**Comment by millzpaugh** [comment]
Off the bat, I'm noticing mbtiles version update issue in upstream dependencies. 

Bumping `mbtiles` to `0.10.0` causes duplicate version installs on this library in unpacker (and potentially other libraries). 

**unpacker** 
There are several dep libraries in unpacker that only allow patch updates of `mbtiles`, while the latest release for `mapbox-upload-validate` branch is a minor bump from `0.9.0` to `0.10.0`. 

We’ll need to bump these dependency libraries to `0.10.0`, ensure tests are passing and and do new releases. 

These libraries are: 

  - [ ] mapbox-geostats 
  - [ ] mapbox-grid-copy 
  - [ ] mapbox-map-metadata 
  - [ ] mapbox-tile-copy 
  - [ ] preprocessorcer  

**altas** 
In relation in atlas-upload-validate, we’ll need to update: 

- pinned `mbtiles` in `package.json`: https://github.com/mapbox/atlas-upload-validate/blob/master/package.json#L34

**Comment by sumarlidason** [comment]
ref: https://github.com/mapbox/mapbox-geostats/pull/40

**Comment by sumarlidason** [comment]
ref: https://github.com/mapbox/mapbox-grid-copy/pull/15

**Comment by sumarlidason** [comment]
ref: https://github.com/mapbox/mapbox-map-metadata/pull/50

**Comment by sumarlidason** [comment]
ref https://github.com/mapbox/mapbox-tile-copy/pull/137
ref https://github.com/mapbox/mapbox-tile-copy/pull/135

**Comment by sumarlidason** [comment]
ref: https://github.com/mapbox/preprocessorcerer/issues/93

**Comment by millzpaugh** [comment]
@sumarlidason These dependency updates look 💯. Thanks for going above basic needs of this PR and improving all the libraries with ES6 updates, adding codecov, test clean up, etc.. 😁

**Pull request merged**